main/src/ca/uwaterloo/flix/runtime/interpreter/Channel.java

```diff
@@ -0,0 +1,164 @@
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public Object tryGet() {
+    lock.lock();
+
+    try {
+      //checkIfClosed(); you can read from a closed channel
+      return queue.poll();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void checkIfClosed() {
+    if (!isOpen) throw new RuntimeException();
+  }
+
+  public void addGetter(Condition c) {
+    lock.lock();
+    try {
+      waitingGetters.add(c);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void close() {
+    lock.lock();
+    try {
+      checkIfClosed();
+      isOpen = false;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public int getId() {
+    return id;
+  }
+
+  public static SelectChoice select(Channel[] channels) {
+    // Create new Condition and lock the current thread
+    Lock selectLock = new ReentrantLock();
+    Condition condition = selectLock.newCondition();
+    selectLock.lock();
+
+    // Sort Channels to avoid deadlock when locking
+    Channel[] sortedChannels = sortChannels(channels);
+
+    while (!Thread.interrupted()) {
+      // Lock all Channels in sorted order
+      lockAllChannels(sortedChannels);
+
+      try {
+        // Check if any Channel has an element
+        for (int index = 0; index < channels.length; index++) {
+          Channel channel = channels[index];
+          Object element = channel.tryGet();
+          if (element != null) {
+            // Element found.
+            // Return the element and the branchNumber (index) of the containing Channel
+            SelectChoice choice = new SelectChoice();
+            choice.branchNumber = index;
+            choice.element = element;
+            return choice;
+          }
+        }
+
+        // No element was found.
+        // Add our condition to all Channels to get notified when a new element is added
+        for (Channel channel : channels) {
+          channel.addGetter(condition);
+        }
+      } finally {
+        // Unlock all Channels in sorted order, so other threads may input elements
+        unlockAllChannels(sortedChannels);
+      }
+
+      // Wait for an element to be added to any of the Channels
+      try {
+        condition.await();
+      } catch (InterruptedException e) {
+        throw new RuntimeException("Thread interrupted");
+      }
+    }
+
+    throw new RuntimeException("Thread interrupted");
+  }
+
+  private static Channel[] sortChannels(Channel[] channels) {
+    Channel[] channelsCopy = Arrays.copyOf(channels, channels.length);
+    Arrays.sort(channelsCopy, Comparator.comparing((Channel c) -> c.id));
+    return channelsCopy;
+  }
+
+  private static void lockAllChannels(Channel[] channels) {
+    // Arrays.sort(channels, Comparator.comparing((Channel c) -> c.id));
+    for (Channel c : channels) c.lock.lock();
+  }
+
+  private static void unlockAllChannels(Channel[] channels) {
+    for (Channel c : channels) {
+      c.lock.unlock();
+    }
+  }
+}
```

Comment:
We should see if CyclicBarrier or CountDownLatch could be use to implement some of this functionality. I don't believe so, but it would be good to know for sure 100%.